### DISCUSSION TITLE
Initial testing error and questions about inference on new videos

### DISCUSSION THREAD

**Discussion by camilletestard** [Skovorp/feral#6] [General]
Hi Peter and Jacopo! I was very excited to see your model come out and already tried it on my video data! I encountered a few bugs when trying to run on colab that I though I could highlight here: 

1. It seems that your code that takes as input Boris .tsv files is maybe not up to date with the latest Boris version? I systematically got empty behavior detections. So i edited that code to make it compatible with the format I am working with (v. 9.6.5 2025-09-02, code attached
[convert_boris_to_feral.py](https://github.com/user-attachments/files/23664444/convert_boris_to_feral.py)
). 

2. It seems that your run.py pipeline expects 'is_multilabel' to be in the labels_json. It currently isn't and adding it as an output somehow didn't solve the problem. As a quick fix I added the following to train.py: 
_def main(cfg):
    with open(cfg['data']['label_json'], 'r') as f:
        labels_json = json.load(f)
    class_names = {int(x): y for x, y in labels_json['class_names'].items()}
    num_classes = len(class_names)

    **if 'is_multilabel' not in labels_json: 
      labels_json['is_multilabel'] = False**_

3. I tried to install your pipeline on Windows and it is not compatible with your Triton pipeline which requires Linux. Training without this on windows seems prohibitively slow -maybe worth mentioning on your page. 

With these two fixes I was able to run the model on Google Colab using a university account to access the A100 GPU. I have yet to see the output (it's currently running).

I have some questions for you moving forward: 
1. As you mention in your notebook, Golab will crash after ~8hours runtime. Can we continue training the model from a checkpoint? 
2. How do I run inferences on new videos from a trained model?
3. How many labeled frames / class do you think users should budget for an initial 
4. Are you planning on setting up guidelines for use on a cluster? I can imagine that would be preferable for more users at scale since those are available resources in many universities... Just a thought. 

Thanks for this wonderful contribution!

**Comment by JacopoRazzauti** [comment]
Hi Camille! 

Thank you for trying FERAL so quickly (and starting our first discussion thread ever!). Indeed, as we just released FERAL we expect to have some bugs to fix in the next few weeks (and realistically in the next few year?), so thank you for flagging some of them already!

1. Yes, our conversion script did not match the latest BORIS output format. Apparently, in the newer version the Comment column is where the START and STOP are reported for the various behaviors (it used to be Status). The new colab version now has a script inspired by the one you shared with us that can flexibly manage different column names.
2. Also true, we recently changed the json format to support multi-label annotations but forgot to update the docs. In your case, setting it to False was correct. We fixed our website, now you can find details on is_multilabel in the Label Preparation tab, and the Dataset Validator is up to date. Like this:

```json
{
  "is_multilabel": true,
  "class_names": {
    "0": "contact",
    "1": "attempt",
    "2": "flight",
    "3": "kicking",
    "4": "genital_contact",
    "5": "mating"
  },
  "labels": {
    "035FQHMNRk.mp4": [
      [0,0,1,0,0,0],
      [1,0,0,0,0,0], # this means that behavior indexed as 0, in this case, contact is happening
      [1,1,0,0,0,0], # this means that both contact and attempt are happening in this single frame
      [0,0,0,0,0,0]
    ]
  },
  "splits": {
    "train": ["035FQHMNRk.mp4"],
    "val": [],
    "test": [],
    "inference": ["035FQHMNRk.mp4"]
  }
}
```

We have now changed the script in Colab to reflect these two cases, and we have also updated the label preparation page on our website to make this clearer.

Your run has properly finished, you can find the results here: https://wandb.ai/sposiboh/feral_public/runs/65k793tm?
Whenever you start a training, you will be prompted to choose open or private as a running mode. If you choose open your run will be public on our public wandb, as it is in this case. The results seem promising so we would be happy to chat about them with you, if you want you can send us an email!

Before running the training, you can check your class distribution across videos and across splits using the dataset validator tool on our website. We find this very useful when assessing our dataset before running the training! This will show you which classes are common and which are rare so you can add more label or exclude them.

Questions:

**- As you mention in your notebook, Golab will crash after ~8hours runtime. Can we continue training the model from a checkpoint?
- How do I run inferences on new videos from a trained model?
- How many labeled frames / class do you think users should budget for an initial
- Are you planning on setting up guidelines for use on a cluster? I can imagine that would be preferable for more users at scale since those are available resources in many universities... Just a thought.
-** 

1 and 2: checkpoints are stored in feral/checkpoints during and after a run and you can already run inference on pretrained checkpoint! To do so you just have to do run.py video_folder label_json_path --checkpoint pth/to/checkpoint  
3. The budget of labels depends on how "subtle" the behavior is. Figure 2g and 4h in our preprint https://www.biorxiv.org/content/10.1101/2025.11.16.688666v1.full show that FERAL is data efficient, but this data efficiency really depends on how visually distinguishable the behavior is in the frame. So we suggest having more labels for classes that are visually subtle. The dataset validator tool will flag classes that are rare in your training dataset, so you can add more labels or exclude them temporarily from the training.
4. We have had success with installing FERAL on different clusters of different universities, so it is totally doable, and you can follow the same instructions as the local installation, or you can use the docker_test file in our repo. But I think, before proceeding maybe it would be good to have a zoom meeting to check if FERAL can run on your cluster. Right now you are running FERAL locally on a 	NVIDIA GeForce RTX 4070 which has 12gB or RAM, this is currently too small for FERAL to run optimally so the cluster might be the best solution!

Hope that this helps and please feel free to reach out via email if you want to discuss your run results and the cluster setup!

Jacopo and Peter

**Reply by camilletestard** [reply]
Wow thanks for such a fast and thorough response! I will let you know how things progress and shoot you an email or post here if i have further issues/questions.
